### PR TITLE
fencinglanguage reads from code and/or pre tags, with test

### PR DESCRIPTION
**Pull request by jywarren**
Hope this looks good to you; it passes the new test and is fairly compact. Open to suggestions of course!

### CODE REVIEW

**Comment by jywarren** [comment]
argh, forgot to build; updating in a moment

**Comment by jywarren** [comment]
built

**Comment by bevacqua** [comment]
Builds should only be included in deployments, and `npm run deploy` takes care of that, so please don't include changes to `dist` :)

File: domador.js [hunk 1]
```diff
@@ -490,7 +490,11 @@ Domador.prototype.process = function process (el) {
     case 'PRE':
       if (this.options.fencing) {
         this.append('\n\n');
-        this.output('```' + (this.options.fencinglanguage(el) || '') + '\n');
+        var fencinglanguage = this.options.fencinglanguage(el);
```
Comment (bevacqua) [inline domador.js]: Let's extract this whole logic into a separate function, as it's getting out of hand.

File: domador.js [hunk 2]
```diff
@@ -490,7 +490,11 @@ Domador.prototype.process = function process (el) {
...
+        if (el.childNodes[0].tagName === 'CODE') {
```
Comment (bevacqua) [inline domador.js]: This will throw if `el` has no child nodes

File: domador.js (hunk 1 above)
Comment (jywarren) [inline domador.js]: Cool, say, everything between 491-499 could become `domador.languagefence()`, or what would be your preferred function name?

Comment (bevacqua) [inline domador.js]: `this.openCodeFence()` and `this.closeCodeFence()` could work.

**Comment by jywarren** [comment]
OK, I made `openCodeFence`, but do you really think `closeCodeFence` is worth it for only one line, `this.outputLater('\n```');` ? I can commit that too if so, for symmetry.

File: domador.js [hunk 3]
```diff
@@ -659,4 +659,12 @@ Domador.prototype.isVisible = function isVisible (el) {
   return visible;
 };
 
+Domador.prototype.openCodeFence = function openCodeFence (el) {
+  var fencinglanguage = this.options.fencinglanguage(el);
+  if (el.childNodes.length > 0 && el.childNodes[0].tagName === 'CODE') {
```
Comment (bevacqua) [inline domador.js]: You could do just `var child = el.childNodes[0]; if (el && el.tagName === 'CODE')`

**Comment by bevacqua** [comment]
Great. So close!

File: domador.js (hunk 3 above)
Comment (jywarren) [inline domador.js]: wait, don't you mean `var child = el.childNodes[0]; if (el && el.childNodes && el.childNodes[0] && el.childNodes[0].tagName === 'CODE')` ? Not that that's elegant at all, but wouldn't `el.childNodes[0]` still throw an error if the el has no childNodes?

Thanks for the tips!

Comment (bevacqua) [inline domador.js]: Meant `var child = el.childNodes[0]; if (child && child.tagName === 'CODE')`

Comment (jywarren) [inline domador.js]: ah, perfect. Pushing updated commits now.

**Comment by jywarren** [comment]
Updated -- all tests passing and /dist/ content removed. Hopefully this is it, thanks for your patience!

File: domador.js [hunk 4]
```diff
@@ -659,4 +659,13 @@ Domador.prototype.isVisible = function isVisible (el) {
   return visible;
 };
 
+Domador.prototype.openCodeFence = function openCodeFence (el) {
+  var fencinglanguage = this.options.fencinglanguage(el);
+  var child = el.childNodes[0];
+  if (child && child.tagName === 'CODE') {
+    fencinglanguage = fencinglanguage || this.options.fencinglanguage(el.childNodes[0]);
```
Comment (bevacqua) [inline domador.js]: You can move the || check to the if above

Comment (jywarren) [inline domador.js]: hmm, internal "code" fencinglanguage should override parent element fencinglanguage, right?

Sorry, I'm not sure I follow... can't think of a way to move the || check into line 665's conditional.

Comment (bevacqua) [inline domador.js]: `if (!fencinglanguage && child && child.tagName === 'CODE') {`

**Comment by jywarren** [comment]
Great, incorporated this and it passes tests. Just a note to self that code language defined on inner blocks is overruled by that defined by outer blocks.

**Comment by bevacqua** [comment]
Great. Thanks for all your work here. `2.4.2`

**Comment by jywarren** [comment]
Thanks for your patience and code review. Much appreciated!

**Pull request merged**